lib/sensors/src/URM04Sensor.cpp

```diff
@@ -0,0 +1,189 @@
+    : _trigPin(trig_pin), startAddr(0x11), readingStep(0), RX(_RX), TX(_TX) {
+  // start timer
+  clock.start();
+  // get start time
+  managerTimer = clock.read_ms();
+
+  // write low to pin to start instructions
+  _trigPin.write(0);
+
+  // Initialize urm04 command recieving address
+  for (int i{0}; i < urmAccount; i++) {
+    urmID[i]   = startAddr + 1;
+    urmData[i] = 0;
+  }
+
+  // initialise commands - set zero everywhere
+  for (int i{0}; i < 10; i++) cmdst[i] = 0;
+
+  // initialize urm04 protocol
+  cmdst[0] = 0x55;
+  cmdst[1] = 0xaa;
+
+  // set nucleo_board baud rate
+
+  // set up serial communication uart bufferserial with baud rate
+  // serial communication - D1/TX , D0/RX,
+  BufferedSerial nucleo_board(TX, RX, BUAD_RATE);
+}
+// destructor
+URM04Sensor::~URM04Sensor() {
+  // kill the timer
+  clock.stop();
+}
+// trigger the measurements
+void URM04Sensor::urmTrigger(int id) {
+  // fill the command buffer with trigger commands
+  cmdst[2] = id;
+  cmdst[3] = 0x00;
+  cmdst[4] = 0x01;
+  // send command over serial
+  transmitCommands();
+}
+// reads the distance from URM04
+void URM04Sensor::urmReader(int id) {
+  // fill command buffer with read commands
+  cmdst[2] = id;
+  cmdst[3] = 0x00;
+  cmdst[4] = 0x02;
+  // send command over serial
+  transmitCommands();
+}
+// transmit commands
+void URM04Sensor::transmitCommands() {
+  cmdst[5] = cmdst[0] + cmdst[2] + cmdst[3] + cmdst[4];
+  // delay for 1 second using thread  1000ms = 1sec
+  ThisThread::sleep_for(1000);
+  // send command over serial connection with BufferedSerial write and put into cmdst
+  nucleo_board.write(&cmdst[0], sizeof(cmdst));
+  // delay for 2 seconds 2000ms = 2sec
+  ThisThread::sleep_for(2000);
+}
+
+// this function actually gets the data
+void URM04Sensor::analyzeUrmData(uint8_t cmd[]) {
+  uint8_t checkSum = 0;
+  // add each byte in the command array to create checksum
+  for (int i{0}; i < 7; i++) checkSum += cmd[i];
+  // check is checksum is correct and that the commands are correct as well
+  if (checkSum == cmd[7] && cmd[3] == 2 && cmd[4] == 2) {
+    uint8_t id  = cmd[2] - startAddr;
+    urmData[id] = cmd[5] * 256 + cmd[6];
+  }
+  // else an error occured during transmission
+  else if (cmd[3] == 2 && cmd[4] == 2) {
+    printf("CheckSum Error");
+  }
+}
+
+// decodes URM04 data
+void URM04Sensor::decodeURM04() {
+  // make sure serial is still running
+  if (nucleo_board.readable()) {
+    // timer
+    int timerPoint = clock.read_ms();
+    // counter
+    int RetryCounter{0};
+    // array to hold commands
+    uint8_t cmdrd[10];
+    // fill the array
+    for (int i{0}; i < 10; i++) cmdrd[i] = 0;
+    // start index & indices
+    int indices{0};
+    int start_index{0};
+    // flag to check if the indices in the buffer are valid
+    bool flag{true};
+    // variable to check if the whole buffer is valid
+    bool valid{false};
+    // header bit
+    uint8_t headerNo{0};
+
+    while (RetryCounter < MAX_TRIES && flag) {
```

Comment:
the && flag part is useless. anything && true = anything. why do you need the flag variable?